core/lib/dal/src/consensus_dal.rs

```diff
@@ -480,6 +461,40 @@ impl ConsensusDal<'_, '_> {
             .number
             .map(|number| attester::BatchNumber(number as u64)))
     }
+
+    /// Get the earliest of L1 batch which does not have a corresponding quorum certificate
+    /// and need signatures to be gossiped and collected.
+    ///
+    /// On the main node this means every L1 batch, because we need QC over all of them to be
+    /// able to submit them to L1. Replicas don't necessarily have to have the QC, because once
+    /// the batch is on L1 and it's final, they can get the batch from there and don't need the
+    /// attestations. The caller will have to choose the `min_batch_number` accordingly.
```

Comment:
the logic makes sense but does it need to be in the dal? The fact that we want to apply some limits from below is business logic - so let the calling component apply it. 

So I propose to make this function simpler by not passing the `min` - and applying it on the caller side. Of course this may result in a number for which we already have a signature - but it's OK since the runner skips such blocks. In fact, I would probably go one step further and make the main/EN logic distinction more explicit. On main node I would invoke this function, whereas on EN I would just use the `min_batch_number` without even going to the DB. Why do we do that? To potentially start a couple of batches later? They will be skipped anyway, but we simplify the overall flow a little bit.

Reply:
Let me see if I can explain why I did it this way:

`PersistentBatchStore::earliest_batch_number_to_sign` indeed doesn't take a `min_batch_number` parameter, and it's supposed to return something still worth signing. We discussed that the replica might not have a QC for everything, and although the main node should, for now that won't be guaranteed either, we make best effort. That means once we have a batch number that doesn't have a QC, the `select min(number) from l1_batches where not exists (...)` query will always return the same number, so it's almost surely pointless: when we roll out this feature it will return 0, if that's our first batch. That would mean re-signing all the batches every time, which is wasted because 1) the `BatchVotes` only handles the latest and 2) most likely nobody else is expecting our signature that far back and no QC will form. 

So I wanted to avoid having to re-sign all batches by passing in a lower bound to the query to tell which batches are still interesting. Normally I would choose this number to be the one following the one which has been last submitted to L1 - anything after that might be waiting on a quorum. We don't have that, so I used the latest-10. 

If I returned the minimum without considering what this lower bound is, then as you say we might as well not call it. We could always just re-sign the last 10, regardless of whether they have a quorum or not. You are right that there is nothing terribly wrong with that, since at the moment I didn't do an extra query in the signing loop to check that whether we have a quroum for heights between the minimum and the latest one.

About your comments, I'm not sure we're on the same page:

> may result in a number for which we already have a signature - but it's OK since the runner skips such blocks. 

The `AttesterRunner` currently doesn't skip such blocks, because skipping would require an extra query to check whether they should be skipped. The original logic with the `unsigned_block_numbers` didn't return these blocks, so they were naturally skipped, but now it's just going from the minimum to the latest without skipping.

> On main node I would invoke this function

Theoretically the main node could do the following:
* use the latest-10 logic for best effort, or
* call this function with 0 to make sure everything is always signed and can be submitted, and return that number on an API to any EN who asks what it should sign, 
* look on L1 to see what has already been submitted and call this method with that
* look on L1 and not call it - then there is no need even for an API since everyone sees L1

> whereas on EN I would just use the `min_batch_number` without even going to the DB

Sorry, what value would an EN assign to the `min_batch_number` without going to the DB? Are you talking about the signing loop?

> Why do we do that? To potentially start a couple of batches later? 

So that when a new node starts it only sends attestations to new or pending batches. In practice it means it shouldn't resend any signatures at all (except maybe the last one).

> They will be skipped anyway, but we simplify the overall flow a little bit.

They won't be skipped at the moment. 

But I agree that we can simplify the flow by saying we don't care whatsoever what's in the DB, just look at L1 and sign everything since the last submission. This query is basically standing in for that, since I don't have a way to look at L1 at my disposal, so I delegated it to the `BatchStore`, which I assumed would have this client at some point.

Reply:
Ok, I mostly got it, let’s keep as is for now to see this actually working and then we can revisit.

Even if we don’t skip explititely, it doesn’t hurt to re-sign a batch; and if the logic is always to sign last ten - why not do it explicitly and code exactly that? We may re-sign this way, but that happens anyway as there may be gaps in signatures. But it’s not very important - let’s proceed with testing

Reply:
I mean, the logic is not _supposed_ to be to re-sign the last 10, and it won't, if you restart you node and you just missed 1 batch QC, you won't re-sign 10. 

This allowance for more than the latest is mostly for when you said we might want the batch interval to go down to several seconds instead of a minute, where we discussed we would have to be gathering votes for multiple pending batches concurrently. I would like to get rid of the "last 10" ASAP in favour of something related to L1.

Today if a node is restarted after a longer hiatus than 10 minutes it will re-sign the last 10 blocks, but it will be wasted because even if there was no quorum, the `BatchVotes` will have forgotten everybody else's votes on earlier heights. That will change with BFT-478, and at that point re-signing the last 10 might move the bar in the vote collector and actually produce the missing QCs for all those heights. 

Then, we can replace the "last 10" with an API call to the main node, and finally an API call to L1. 

Once we have the API on the main node, that should be the only one which queries its database to decide which batch it wants signatures for. I thought there might be some flag in there about L1 submissions, but if not, perhaps "last 10" is all it can ever do.

--- 

I think the concept of asking for the first one to sign is correct, we're just unsure where it should be coming from exactly in the in-between stages. This query is just an implementation detail of skipping the prefix of already signed batches, just doing a tiny bit less work. It's not part of any consensus-node interface, so it's pretty easy to remove. I wanted to do the right thing (select the last submission), but because I couldn't, I had to fake it with this "last 10", but that is not _the_ logic, just a cop-out. 

You know what, I'll remove the query and just return the last 10, and leave some more notes about where we want to go with this.

Reply:
Removed.